src/socket.js

```diff
@@ -0,0 +1,57 @@
+let DataReceiver = {
+  callbacks: new Map(),
+  onDataReceived: (aSocketId, aData) => {
+    if (!DataReceiver.callbacks.has(aSocketId)) {
+      return;
+    }
+    DataReceiver.callbacks.get(aSocketId)._updateQueue(aData);
+  },
+};
+
+browser.sockets.onDataReceived.addListener(DataReceiver.onDataReceived);
+
+let ListeningSockets = new Map();
+
+export class ListeningSocket {
+  constructor() {
+    this.queue = []; // stores messages sent to socket
+  }
+
+  async startListening() {
+    this.port = await browser.sockets.createServerSocket();
+    DataReceiver.callbacks.set(this.port, this);
+    browser.sockets.startListening(this.port);
+    console.log('Listening on port ' + this.port);
+  }
+
+  _updateQueue(data) {
+    console.log('_updateQueue data', data);
+    this.queue.push(data);
+  }
+}
+
+export class SendingSocket {
+  constructor() {
+  }
+
+  async connect(host, port) {
+    this.id = await browser.sockets.createSendingSocket();
+    browser.sockets.connect(this.id, host, port);
+    console.log(`Connected to ${host}:${port}`);
+  }
+
+  send(aData, serializationSymbol) {
```

Comment:
We should hide the exact serialization type from the callers here, since it requires them to know the socket header format. I suggest updating this to be `send(aData, aSerialization='json')`, and include a comment that details the supported serialization types. We can then map these human readable types to the symbols in the Extension API.

Reply:
I'm suggesting the default value of `json` since that appears to be what we use throughout. That will let us remove the `'j'` parameters from everything in `loggingdb.js`.

Reply:
Ha, digging into this and the switch statement confusion reveals that we are doing `JSON.parse` on received data before passing it to the onDataReceived listeners - very fishy since they are supposed to receive strings! I'm on it.

Reply:
This also reveals the problem with visit_id